File: Pset 1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool checkCard(long card)
{
    int sum = 0;
    long digit = 100;
    int digitVal = 0;

    // sum of every other digit starting from second to last
    while (card / digit != 0)
    {
        // get the digit value
        digitVal = (card % digit) / (digit / 10);
        digitVal *= 2;
        //printf("This is digit: %i\n", digitVal);

        // double digit catch
        if (digitVal >= 10)
        {
            sum += 1 + (digitVal % 10);
        }
        else
        {
            sum += digitVal;
        }

        // next digit
        digit *= 100;
    }

    // gets last digit
    digitVal = (card % digit) / (digit / 10);
    digitVal *= 2;
    //printf("This is digit: %i\n", digitVal);
    if (digitVal >= 10)
    {
        sum += 1 + (digitVal % 10);
    }
    else
    {
        sum += digitVal;
    }

    //printf("%i\n", sum);

    // gets the remaining digits starting from last on the right
    digitVal = 0;
    digit = 10;
    while (card / digit != 0)
    {
        digitVal = (card % digit) / (digit / 10);
        //printf("This is digit: %i\n", digitVal);

        sum += digitVal;

        digit *= 100;

    }
    // catch last digit
    digitVal = (card % digit) / (digit / 10);
    //printf("This is digit: %i\n", digitVal);
    sum += digitVal;

    return (sum % 10 == 0);
}

int main(void)
{
    // get credit card number
    long cardNum = get_long("Number: ");
    if (cardNum > 5599999999999999)
    {
        printf("INVALID\n");
    }
    else if ((cardNum / 5500000000000000 != 0 || cardNum / 5400000000000000 != 0 || cardNum / 5300000000000000 != 0 ||
              cardNum / 5200000000000000 != 0 || cardNum / 5100000000000000 != 0) && checkCard(cardNum))
    {
        printf("MASTERCARD\n");
    }
    else if ((cardNum / 4000000000000000 != 0) && checkCard(cardNum))
    {
        printf("VISA\n");
    }
    else if (cardNum / 370000000000000 != 0 && checkCard(cardNum))
    {
        printf("AMEX\n");
    }
    else if ((cardNum < 370000000000000 && cardNum > 340000000000000))
    {
        printf("INVALID\n");
    }
    else if (cardNum / 340000000000000 != 0 && checkCard(cardNum))
    {
        printf("AMEX\n");
    }
    else if ((cardNum / 4000000000000 != 0) && checkCard(cardNum))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Week 6/sentimental-credit/credit.py
# TODO
import cs50


def checkValid(number):
    try:
        return isinstance(int(number), int)
    except ValueError:
        return False


def checkDigits(number):
    sum = 0
    tmp = ""

    # Double every other digit starting from second to last
    for i in range(len(number)-2, -1, -2):
        tmp += str(2*int(number[i]))
    # Sum of those digits together not the numbers
    for i in range(len(tmp)):
        sum += int(tmp[i])

    # Sum of those digits that weren't already counted for
    for i in range(len(number)-1, -1, -2):
        sum += int(number[i])

    # Return true if the last digit in the number was 0
    if sum % 10 == 0:
        return True
    else:
        return False


def checkCardType(number):
    # If the first two numbers are [set] and length matches and the digits are valid
    if number[:2] in ["34", "37"] and len(number) == 15 and checkDigits(number):
        print("AMEX")
    elif number[:2] in ["51", "52", "53", "54", "55"] and len(number) == 16 and checkDigits(number):
        print("MASTERCARD")
    elif number[:1] in ["4"] and (len(number) == 16 or len(number) == 13) and checkDigits(number):
        print("VISA")
    else:
        print("INVALID")


def main():
    # Keeps trying to take input until numbers are input
    while True:
        card_num = cs50.get_string("Number: ")
        # If card has only valid inputs then it gets checked
        if checkValid(card_num):
            break
    checkCardType(card_num)


main()